src/osd/PrimaryLogPG.cc

```diff
@@ -7205,17 +7206,24 @@ void PrimaryLogPG::write_update_size_and_usage(object_stat_sum_t& delta_stats, o
     }
     oi.size = new_size;
   }
-  if (length && oi.has_extents()) {
-     delta_stats.num_bytes -= oi.extents.size();
-     if (write_full) {
-       // oi.size may shrink
-       oi.extents.clear();
-       assert(offset == 0);
-       oi.extents.insert(0, length);
-     } else {
-       oi.extents.union_of(ch); // deduplicated
-     }
-     delta_stats.num_bytes += oi.extents.size();
+  if (oi.has_extents()) {
+    delta_stats.num_bytes -= oi.extents.size();
+    if (write_full) {
+      // oi.size may shrink
+      oi.extents.clear();
+      assert(offset == 0);
+      if (length) {
+        oi.extents.insert(0, length);
+      }
+    } else {
+      if (length) {
+        oi.extents.union_of(ch); // deduplicated
+      } else if (offset > old_size) {
+        // equivalent to truncate-up
+        oi.extents.insert(old_size, offset - old_size);
+      }
+    }
+    delta_stats.num_bytes += oi.extents.size();
```

Comment:
I would be inclined not to do this patch since it doesn't change the allocations...

Reply:
@liewegas Okay, updated.